streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -0,0 +1,398 @@
+            final V value = segmentMapEntry.getValue().get(new WrappedK<>(key, seqnum));
+            if (value != null) {
+                returnSet.add(new KeyValue<>(segmentMapEntry.getKey(), value));
+            }
+        }
+        return returnSet;
+    }
+
+    private List<KeyValue<Long, V>> fetchWithDuplicates(final K key, final long timeFrom, final long timeTo) {
+        final List<KeyValue<Long, V>> returnSet = new LinkedList<>();
+
+        // add one b/c records expire exactly retentionPeriod ms after created
+        final long minTime = max(timeFrom, this.context.streamTime() - this.retentionPeriod + 1);
+        final WrappedK<K> keyFrom = new WrappedK<>(key, 0);
+        final WrappedK<K> keyTo = new WrappedK<>(key, Integer.MAX_VALUE);
+
+        for (final Map.Entry<Long, NavigableMap<WrappedK<K>, V>> segmentMapEntry : this.segmentMap.subMap(minTime, true, timeTo, true).entrySet()) {
+            for (final Map.Entry<WrappedK<K>, V> kvMapEntry : segmentMapEntry.getValue().subMap(keyFrom, true, keyTo, true).entrySet()) {
+                returnSet.add(new KeyValue<>(segmentMapEntry.getKey(), kvMapEntry.getValue()));
+            }
+        }
+        return returnSet;
+    }
+
+    private void removeExpiredSegments() {
+        final long minLiveTime = this.context.streamTime() - this.retentionPeriod;
+        final NavigableMap<Long, NavigableMap<WrappedK<K>, V>> expiredSegments = this.segmentMap.headMap(minLiveTime, true);
+        for (Iterator<Entry<Long, NavigableMap<WrappedK<K>, V>>> it = expiredSegments.entrySet().iterator(); it.hasNext(); ) {
+            it.next();
+            it.remove();
+        }
+    }
+
+    private KeyValue<Windowed<K>, V> getWindowedKeyValue(final K key, final long startTimestamp, final V value) {
+        final Windowed<K> windowedK = new Windowed<>(key, new TimeWindow(startTimestamp, startTimestamp + windowSize));
+        return new KeyValue<>(windowedK, value);
+    }
+
+    private void maybeUpdateSeqnumForDups() {
+        if (retainDuplicates) {
+            seqnum = (seqnum + 1) & 0x7FFFFFFF;
+        }
+    }
+
+    private static class WrappedK<K extends Comparable<K>> implements Comparable<WrappedK<K>> {
+        private final K key;
+        private final int seqnum;
+
+        WrappedK(final K key, final int seqnum) {
+            this.key = key;
+            this.seqnum = seqnum;
+        }
+
+        public K getKey() {
+            return this.key;
+        }
+
+        public int compareTo(final WrappedK<K> k) {
+            final int compareKeys = this.key.compareTo(k.key);
+            if (compareKeys == 0) {
+                return this.seqnum - k.seqnum;
+            } else {
+                return compareKeys;
+            }
+        }
+    }
+
+    private static class InMemoryWindowStoreIterator<V> implements WindowStoreIterator<V> {
```

Comment:
I think we can consolidate these two iterators and the `InMemoryKeyValueStore#InMemoryKeyValueIterator` as a single wrapper iterator (note `WindowStoreIterator<V>` is just a `KeyValueStoreIterator<Long, V>.), that takes an `iterator` as inner. Then for supporting "peekNextKey" we do not really need to rely on list iterator's `previous` but just need to maintain the next element, and update it whenever `next()` is called and this element on-hold is returned.

And when we removed the generics we can even consolidate all of them in `WindowStoreIteratorWrapper` as well.

Reply:
The problem with using the WindowStoreIteratorWrapper here is that expects the key Bytes to actually be a windowed key ie to contain the timestamp, however with the in-memory version we want to avoid having to wrap the timestamp with the key as well so the iterator cant just extract it from the Bytes

Reply:
When we removed the generics and allow have bytes in `InMemoryWindowStore`, when we return the inner ListIterator aren't we going to have each key be concatenated with key-bytes followed with timestamp-bytes anyways?

Reply:
Currently the key Bytes are wrapped up with timestamp into a windowed<key> type directly in the fetch call, whereas the WindowStoreIteratorWrapper is given an iterator over Bytes and it does the extraction of key+timestamp -> windowed<key> when accessed. We could rewrite the in-memory store as a single map with key+timestamp bytes as the key, but with the current implementation wrapping the key with the timestamp is a waste of space as we already map from each timestamp to a kv store

Reply:
John actually raised that as a possibility to improve clarity. I wasn't entirely convinced that it would make things easier to read but if you both agree, I can be. Certainly I agree it would be good to share the iterator code

Reply:
In retrospect, I'm not sure I'm convinced either ;)

One nice aspect of your current design is that when we need to drop a window, we can do it in one operation.

Reply:
Okay, I think I'm convinced. We should just leave it as is in this PR at least, and after it is merged I will think about it twice when reviewing the other PR for inlining the byte stores.